File: bindings/python/kevy/_resp_conn.py
"""Blocking RESP2/RESP3 TCP transport (contract §1.1).

One request at a time; a read buffer reassembles multi-segment replies. A
clean server close mid-read surfaces as ``ClosedError``; a socket read
timeout as ``TimedOutError``; any other socket failure as ``IoError``. Not
safe for concurrent use from multiple threads.
"""

from __future__ import annotations

import socket
from typing import List, Optional

from ._errors import ClosedError, IoError, ProtocolError, TimedOutError
from ._reply import Reply, parse_reply


def encode_command(argv: List[bytes]) -> bytes:
    parts = [b"*", str(len(argv)).encode(), b"\r\n"]
    for a in argv:
        parts.extend([b"$", str(len(a)).encode(), b"\r\n", a, b"\r\n"])
    return b"".join(parts)


class RespConn:
    def __init__(self, sock: socket.socket, host: str, port: int):
        self._sock: Optional[socket.socket] = sock
        self._rbuf = bytearray()
        self.host = host
        self.port = port

    @classmethod
    def dial(cls, host: str, port: int) -> "RespConn":
        try:
            sock = socket.create_connection((host, port))
            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        except OSError as e:
            raise IoError(e)
        return cls(sock, host, port)

    def close(self) -> None:
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
            self._sock = None

    @property
    def closed(self) -> bool:
        return self._sock is None

    def request(self, argv: List[bytes]) -> Reply:
        if self._sock is None:
            raise ClosedError()
        self._write_all(encode_command(argv))
        return self._read_one()

    def pipeline_raw(self, wire: bytes, n: int) -> List[Reply]:
        if self._sock is None:
            raise ClosedError()
        self._write_all(wire)
        out: List[Reply] = []
        while len(out) < n:
            out.append(self._read_one())
        return out

    def write(self, argv: List[bytes]) -> None:
        """Send one command without reading a reply — the decoupled send
        side used by the pub/sub consumer."""
        if self._sock is None:
            raise ClosedError()
        self._write_all(encode_command(argv))

    def read_reply(self) -> Reply:
        """Block for the next reply frame (used by the pub/sub consumer)."""
        if self._sock is None:
            raise ClosedError()
        return self._read_one()

    def set_read_timeout(self, seconds: Optional[float]) -> None:
        if self._sock is None:
            raise ClosedError()
        self._sock.settimeout(seconds)

    def select_db(self, db: int) -> None:
        r = self.request([b"SELECT", str(int(db)).encode()])
        if r.is_error():
            raise ProtocolError(f"SELECT {db} rejected: {r.text()}")

    # --- internals ------------------------------------------------------

    def _write_all(self, wire: bytes) -> None:
        assert self._sock is not None
        try:
            self._sock.sendall(wire)
        except socket.timeout:
            raise TimedOutError()
        except OSError as e:
            self.close()
            raise IoError(e)

    def _read_one(self) -> Reply:
        assert self._sock is not None
        while True:
            try:
                reply, used = parse_reply(bytes(self._rbuf))
            except ProtocolError:
                raise ProtocolError("malformed reply")
            if used > 0:
                del self._rbuf[:used]
                return reply  # type: ignore[return-value]
            try:
                chunk = self._sock.recv(8192)
            except socket.timeout:
                raise TimedOutError()
            except OSError as e:
                self.close()
                raise IoError(e)
            if not chunk:
                self.close()
                raise ClosedError()
            self._rbuf.extend(chunk)

File: bindings/java/src/main/java/jp/golia/kevy/RespConn.java
// RespConn — the blocking RESP2/3 TCP transport (client-contract §1.1 remote).
// A hand-rolled socket client: encode a command's argv, write it, and read
// replies back with the shared incremental parser feeding a growable buffer
// (the exact "parse → need-more → read more" loop the other ports use). Used
// by the command connection, the Subscriber, the pipeline, and each cluster
// shard.
package jp.golia.kevy;

import java.io.BufferedOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

final class RespConn implements AutoCloseable {
    private final Socket sock;
    private final InputStream in;
    private final OutputStream out;
    private byte[] rbuf = new byte[8192];
    private int rstart = 0;
    private int rend = 0;

    private RespConn(Socket sock) throws IOException {
        this.sock = sock;
        this.in = sock.getInputStream();
        this.out = new BufferedOutputStream(sock.getOutputStream());
    }

    static RespConn dial(String host, int port) {
        try {
            Socket s = new Socket();
            s.setTcpNoDelay(true);
            s.connect(new InetSocketAddress(host, port), 10_000);
            return new RespConn(s);
        } catch (IOException e) {
            throw new KevyIoException("connect to " + host + ":" + port + " failed", e);
        }
    }

    /** Send a command and read exactly one reply. */
    Reply request(List<byte[]> argv) {
        writeCommand(argv);
        flush();
        return readOne();
    }

    /** Write a command without reading its reply (Subscriber subscribe/etc.). */
    void writeOnly(List<byte[]> argv) {
        writeCommand(argv);
        flush();
    }

    /** Send a pre-encoded batch of `count` commands and read `count` replies. */
    List<Reply> pipelineRaw(byte[] wire, int count) {
        try {
            out.write(wire);
        } catch (IOException e) {
            throw mapIo(e);
        }
        flush();
        List<Reply> replies = new ArrayList<>(count);
        for (int i = 0; i < count; i++) replies.add(readOne());
        return replies;
    }

    /** Issue SELECT N on connect; a -ERR reply fails the connect. */
    void selectDb(int db) {
        Reply r = request(List.of(Bytes.of("SELECT"), Bytes.ofLong(db)));
        if (r.isError()) throw Errors.fromReplyText(r.payload());
    }

    /** Bound blocking reads; null clears the deadline. */
    void setReadTimeout(java.time.Duration dur) {
        try {
            sock.setSoTimeout(dur == null ? 0 : (int) Math.max(1, dur.toMillis()));
        } catch (IOException e) {
            throw mapIo(e);
        }
    }

    private void writeCommand(List<byte[]> argv) {
        try {
            RespParser.encodeCommand(out, argv); // straight into the buffered stream, no intermediate byte[]
        } catch (IOException e) {
            throw mapIo(e);
        }
    }

    private void flush() {
        try {
            out.flush();
        } catch (IOException e) {
            throw mapIo(e);
        }
    }

    /** Read one full reply, refilling the buffer from the socket as needed. */
    Reply readOne() {
        while (true) {
            RespParser.Parsed p = RespParser.parse(rbuf, rstart, rend);
            if (p.used() > 0) {
                rstart += p.used();
                return p.reply();
            }
            fill();
        }
    }

    private void fill() {
        if (rstart > 0) {
            System.arraycopy(rbuf, rstart, rbuf, 0, rend - rstart);
            rend -= rstart;
            rstart = 0;
        }
        if (rend == rbuf.length) {
            rbuf = Arrays.copyOf(rbuf, rbuf.length * 2);
        }
        int n;
        try {
            n = in.read(rbuf, rend, rbuf.length - rend);
        } catch (SocketTimeoutException e) {
            throw new TimedOutException("read timed out");
        } catch (IOException e) {
            throw mapIo(e);
        }
        if (n < 0) {
            closeQuietly();
            throw new ClosedException("server closed connection mid-read");
        }
        rend += n;
    }

    private KevyException mapIo(IOException e) {
        return new KevyIoException(e.getMessage() == null ? e.toString() : e.getMessage(), e);
    }

    private void closeQuietly() {
        try {
            sock.close();
        } catch (IOException ignored) {
            // best effort
        }
    }

    @Override
    public void close() {
        closeQuietly();
    }
}
